### PR TITLE
Issue 42. Get field data using the get_data method

### PR DESCRIPTION
**Pull request by sergisiso**
This PR allows OpenCL implementations to provide checksums for any value at any time.

Also will make MPI+OpenCL possible if we ever want to try this route.

### CODE REVIEW

**Comment by sergisiso** [comment]
@arporter PR ready, it is just 2 lines. It works well with all nemolite2d versions but I have not tested OpenACC.

File: finite_difference/src/field_mod.f90 [hunk 1]
```diff
@@ -1009,7 +1009,7 @@ function fld_checksum(field) result(val)
     type(r2d_field), intent(in) :: field
     real(go_wp) :: val
 
-    val = array_checksum(field%data, field%data_on_device,           &
+    val = array_checksum(field%get_data(), field%data_on_device,           &
```
Comment (arporter) [inline finite_difference/src/field_mod.f90:1013]: Please could you delete the commented-out code on lines 1017-1028 as it is no longer necessary.

Comment (arporter) [reply finite_difference/src/field_mod.f90:1013]: Please could you also update the docstring to say that if the data for the field is on a remote device then it is fetched first.

File: finite_difference/src/field_mod.f90 [hunk 2]
```diff
@@ -1040,7 +1040,7 @@ subroutine halo_exchange(self, depth)
     ! Locals
     integer :: exch  !> Handle for exchange
 
-    call exchange_generic(b2=self%data, handle=exch, &
+    call exchange_generic(b2=self%get_data(), handle=exch, &
```
Comment (arporter) [inline finite_difference/src/field_mod.f90:1034]: Please could you similarly update the docstring for this routine to say that data is first copied to the host (if necessary).

**Review by arporter — CHANGES_REQUESTED** [review]
All looks fine. When you say it works for all nemolite2d versions, presumably that means you've tested with OpenCL. If so, that's great :-)
Just a couple of minor tweaks required and then I'll merge it.

File: finite_difference/src/field_mod.f90 (hunk 1 above)
Comment (sergisiso) [reply finite_difference/src/field_mod.f90:1013]: Done

File: finite_difference/src/field_mod.f90 (hunk 2 above)
Comment (sergisiso) [reply finite_difference/src/field_mod.f90:1034]: Done

**Comment by sergisiso** [comment]
@arporter Made requested changes.

> When you say it works for all nemolite2d versions, presumably that means you've tested with OpenCL.

Yes, I also checked that I didn't break the MPI by updating the halo_exchange call.

I still have problems with OpenCL JIT with large tests that segfault from time to time, but this are unrelated to this issue.

**Review by arporter — APPROVED** [review]
All requested changes have been made.

**Pull request merged**